Move project metadata & build specs from setup.py to pyproject.toml

zaneselvans opened the pull request:
# PR Overview

I don't know if this was new with Python 3.11, but I've been getting warnings about our continued use of `setup.py` for packaging, so decided to try and swap it out for the now official `pyproject.toml` build system entirely and I think it's working. Mostly I followed instructions from:
* [Declaring project metadata](https://packaging.python.org/en/latest/specifications/declaring-project-metadata/)
* [Data files support](https://setuptools.pypa.io/en/latest/userguide/datafiles.html)

One thing I'm still not entirely clear on is what files should end up in the source vs. wheel distributions, and how we control that.  We are using `setuptools_scm` which pulls our version from git tags, and also understands which files are under source control, and in theory it's supposed to pull in all "data" (non python) files that it finds under the package directory (`src/pudl` in our case) which is what we want it to do. But when I build a new package by running:

```bash
python -m build
```

I see a whole lot of other files from outside of the package directory getting added to... something. Is that just the [source distribution](https://packaging.python.org/en/latest/specifications/source-distribution-format/) slurping up everything in the whole repository? Is that expected / desired?

Should we get rid of `MANIFEST.in` altogether?

I stuck with `setuptools` as the build backend because I didn't want to change multiple things all at once, but it seems like there are other options with less legacy baggage floating around now, like `flit` and `hatchling` as well as systems that do much more, like `poetry`.

# PR Checklist

- [ ] Merge the most recent version of the branch you are merging into (probably `dev`).
- [ ] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [ ] Make sure you've included good docstrings.
- [ ] For major data coverage & analysis changes, [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
- [ ] Include unit tests for new functions and classes.
- [ ] Defensive data quality/sanity checks in analyses & data processing functions.
- [ ] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html) and reference reference the PR and related issues.
- [ ] Do your own explanatory review of the PR to help the reviewer understand what's going on and identify issues preemptively.

jdangerx commented on MANIFEST.in:
```diff
@@ -1,13 +1,8 @@
 graft src/pudl/package_data
```
If we remove this `graft` line, the package_data is still included because we defined it in `pyproject.toml`.

jdangerx reviewed (approved):
✅ this installs properly with `mamba env create`
✅ `tox -e build` works

I think this is totally reasonable, and if we want to clean up the sdist or the MANIFEST.in by configuring more stuff in `pyproject.toml` we can do it later :)

It appears we do still need the MANIFEST.in so we can get rid of the random `.github`, `.gitattributes`, etc. though I'm not sure why we're excluding our tests. The [default behavior](https://packaging.python.org/en/latest/guides/using-manifest-in/#how-files-are-included-in-an-sdist) appears to include tests, so maybe including them is best practice for source distributions?

Questions that I think aren't *that* important right now:
* why is setuptools including the `.github` repository (and gitattributes, readthedocs.yml, etc.)? They're not listed as default includes in the [docs](https://packaging.python.org/en/latest/guides/using-manifest-in/#how-files-are-included-in-an-sdist).
* should we include the tests or no?

zaneselvans commented:
I think [`setuptools_scm`](https://github.com/pypa/setuptools_scm#file-finders-hook-makes-most-of-manifestin-unnecessary) is probably responsible for pulling EVERYTHING into the source distribution, which is annoying. By default it appears to pull in every file that's tracked by git.

I think we excluded the tests because we didn't intent for them to be run by someone who was just installing the package, and at some point there were a few of them interleaved with the package modules, rather than being 100% isolated under the `test` directory.

We should probably either prune or include the entire `test` directory. Does it make sense to distribute the tests with the source distribution?

Does it make sense to exclude just `.readthedocs.yml` but include literally all the other little configuration files?

What would be the simplest possible criteria for choosing files to exclude from the source distribution, rather than the random mish-mash we've got now, including selectors that don't match any files, and inconsistent exclusion of random files?

zaneselvans commented:
Some possibilities for pruning:
* We have like 50MB of PDFs under the `docs`. Does it make sense to distribute `docs` at all? Shouldn't people go to the documentation itself on RTD, or GitHub?  Also we automatically archive the entire repo on Zenodo every time we tag a release.
* Ditto the `test` directory. I don't understand why it would be reasonable to distribute it in the `sdist` But maybe I don't understand what the `sdist` is for?

jdangerx commented:
Re: the test files, the [docs](https://docs.python.org/3/distutils/sourcedist.html#specifying-the-files-to-distribute) indicate that tests are included because of a "future standard for testing Python distributions."

So there's that. I think the simplest thing would be "we want to distribute exactly the things that are distributed by default."

> - all Python source files implied by the py_modules and packages options
> - all C source files mentioned in the ext_modules or libraries options
> - scripts identified by the scripts option See [Installing Scripts](https://docs.python.org/3/distutils/setupscript.html#distutils-installing-scripts).
> - anything that looks like a test script: test/test*.py (currently, the Distutils don’t do anything with test scripts except include them in source distributions, but in the future there will be a standard for testing Python module distributions)
> - Any of the standard README files (README, README.txt, or README.rst), setup.py (or whatever you called your setup script), and setup.cfg.
> - all files that matches the package_data metadata. See [Installing Package Data](https://docs.python.org/3/distutils/setupscript.html#distutils-installing-package-data).
> - all files that matches the data_files metadata. See [Installing Additional Files](https://docs.python.org/3/distutils/setupscript.html#distutils-additional-files).

I took a look and, yes, setuptools_scm is the culprit for pulling *everything* in from git. Their [recommendation](https://github.com/pypa/setuptools_scm#file-finders-hook-makes-most-of-manifestin-unnecessary) seems to be to exclude a bunch of stuff with MANIFEST.in.

Given all this, I think we should:

* keep MANIFEST.in, and live with the clutter
* aim to have our source distribution match what a "normal" one would look like (keep tests, source, scripts, README, package data, pyproject.toml, ditch everything else)

jdangerx commented:
I see your latest commit prunes out `test. I'm not that invested in keeping it in. When the new standard eventually drops and someone wants the tests in the sdist, either:

* someone will ask us
* they will check out the repo themselves
* we will remember to un-prune the tests dir

Which all seem fine to me.

zaneselvans commented:
I was most ambivalent about removing the tests, so I've added it back in, and removed all the dotfiles / config stuff.

Perhaps not coincidentally, our source and binary distributions are now exactly the same size (2.7MB down from 32MB previously!)

I strongly suspect that "future standard" is never coming. That whole page of documentation is a zombie that's supposed to be taken over by the now distinct `setuptools` and I get the feeling that `setuptools` is also a bit of a zombie. There's a lot of action in Python packagingland these days and hopefully something good will settle out of it.
